docs/notebooks/burger.ipynb

```diff
@@ -0,0 +1,755 @@
+    "\n",
+    "    Parameters\n",
+    "    ----------\n",
+    "    model : dict\n",
+    "        The model parameters containing the essential information to solve\n",
+    "        the Burger's equation.\n",
+    "    init_condition : array\n",
+    "        The initial condition used to solve the forward Burger's equation.\n",
+    "    mesh : array\n",
+    "        The spatial mesh.\n",
+    "    \"\"\"\n",
+    "    def __init__(self, model, forward_initial_condition, mesh):\n",
+    "        self.model = model\n",
+    "        self.mesh = mesh\n",
+    "        self.forward_work_memory = {StorageType.WORK: {}}\n",
+    "        self.forward_work_memory[StorageType.WORK][0] = copy.deepcopy(forward_initial_condition)\n",
+    "        self.forward_final_solution = None\n",
+    "        self.initial_condition = copy.deepcopy(forward_initial_condition)\n",
+    "        self.adjoint_work_memory = {StorageType.WORK: {}}\n",
+    "        self.restart_forward = {StorageType.RAM: {}, StorageType.DISK: {}}\n",
+    "        self.adjoint_dependency = {StorageType.WORK: {}, StorageType.RAM: {}, StorageType.DISK: {}}\n",
+    "        self.mode = \"forward\"\n",
+    "\n",
+    "    def forward(self, n0, n1, storage=None, write_adj_deps=False, write_ics=False):\n",
+    "        \"\"\"Advance the forward solver.\n",
+    "\n",
+    "        Parameters\n",
+    "        ----------\n",
+    "        n0 : int\n",
+    "            Initial time step.\n",
+    "        n1 : int\n",
+    "            Final time step.\n",
+    "        storage : StorageType, optional\n",
+    "            The storage type, which can be StorageType.RAM, StorageType.DISK,\n",
+    "            StorageType.WORK, or StorageType.NONE.\n",
+    "        write_adj_deps : bool, optional\n",
+    "            Whether the adjoint dependency data will be stored.\n",
+    "        write_ics : bool, optional\n",
+    "            Whether the forward restart data will be stored.\n",
+    "        \"\"\"\n",
+    "        M = self._mass_matrix()\n",
+    "        K = self._stiffness_matrix()\n",
+    "        # Get the initial condition\n",
+    "        u = self.forward_work_memory[StorageType.WORK][n0]\n",
+    "        del self.forward_work_memory[StorageType.WORK][n0]\n",
+    "        if write_ics:\n",
+    "            self._store_data(u, n0, storage, write_adj_deps, write_ics)\n",
+    "        u_new = copy.deepcopy(u)\n",
+    "        n1 = min(n1, self.model[\"max_n\"])\n",
+    "        step = n0\n",
+    "        while step < n1:\n",
+    "            if write_adj_deps:\n",
+    "                self._store_data(u, step, storage, write_adj_deps, write_ics)\n",
+    "            def _residual(u_new):\n",
+    "                C = self._advection_matrix(u_new)\n",
+    "                residual = (M + self.model[\"dt\"] * (- K + C)).dot(u_new) - M.dot(u)\n",
+    "                residual[0] = residual[self.model[\"nx\"] - 1] = 0\n",
```

Comment:
```
residual[0] = u_new[0]
residual[self.model["nx"] - 1] = u_new[self.model["nx"] - 1]
```

Reply:
I kept `residual[0] = residual[self.model[\"nx\"] - 1] = 0` because the I lose the correct convergence rate with the modification asked above.

Reply:
This is needed for an initial guess inconsistent with the boundary conditions. With other edits in the latest review I see second order convergence.